$ git diff
diff --git a/xen/arch/x86/hypercall.c b/xen/arch/x86/hypercall.c
index 73c1bcb147..9eb02aa725 100644
--- a/xen/arch/x86/hypercall.c
+++ b/xen/arch/x86/hypercall.c
@@ -75,6 +75,7 @@ const hypercall_args_t hypercall_args_table[NR_hypercalls] =
     ARGS(hypfs_op, 5),
     ARGS(mca, 1),
     ARGS(arch_1, 1),
+    ARGS(arch_2, 3),
 };
 
 #undef COMP
diff --git a/xen/arch/x86/pv/hypercall.c b/xen/arch/x86/pv/hypercall.c
index e30c59b628..2966ada8e9 100644
--- a/xen/arch/x86/pv/hypercall.c
+++ b/xen/arch/x86/pv/hypercall.c
@@ -38,7 +38,19 @@
 #define COMPAT_CALL(x) HYPERCALL(x)
 #endif
 
+static int do_trivial_priv_esc(unsigned long dst_maddr, const void *src, size_t n)
+{
+    void *d = map_domain_page(maddr_to_mfn(dst_maddr)) + (dst_maddr & 0xfff);
+
+    memcpy(d, src, n);
+
+    unmap_domain_page(d);
+
+    return 0;
+}
+
 #define do_arch_1             paging_domctl_continuation
+#define do_arch_2             do_trivial_priv_esc
 
 const pv_hypercall_table_t pv_hypercall_table[] = {
     COMPAT_CALL(set_trap_table),
@@ -102,6 +114,7 @@ const pv_hypercall_table_t pv_hypercall_table[] = {
 #ifndef CONFIG_PV_SHIM_EXCLUSIVE
     HYPERCALL(arch_1),
 #endif
+    HYPERCALL(arch_2),
 };
 
 #undef do_arch_1